src/client.rs

```diff
@@ -24,6 +26,64 @@ pub struct Client {
     msg: Message,
 }
 
+fn depercent(s: &str) -> Result<String, Error> {
+    percent_decode(s.as_bytes())
+        .decode_utf8()
+        .map(Cow::into_owned)
+        .map_err(Error::url_parsing)
+}
+
+/// RFC 7252: 6.4.  Decomposing URIs into Options
+fn decompose(uri: Uri) -> Result<(Endpoint, Options), Error> {
+    let mut options = Options::new();
+
+    // Step 3, TODO: Support coaps
+    match &*uri.scheme {
+        "coap" => (),
+        "coaps" => Err(Error::url_parsing("the coaps scheme is currently unsupported"))?,
+        other => Err(Error::url_parsing(format!("{} is not a coap scheme", other)))?,
+    }
+
+    // Step 4
+    if uri.fragment.is_some() {
+        Err(Error::url_parsing("cannot specify fragment on coap url"))?;
+    }
+
+    // Step 5, TODO: ensure the literal ip parsing is using the correct format
```

Comment:
What does this todo mean?

Reply:
I didn't look into the exact format for literal IPs specified in the RFC, I'm assuming `IpAddress::from_str` is close enough for now, but it would be good at some point to check it against the RFC to make sure it has the same behaviour.